File: sdks/go/filter.go
package vectorizer

// QdrantFilter mirrors the server's filter wire shape (phase23).
//
// All three clause arrays are optional; omit any you don't need. At least
// one clause with at least one condition must be present — the server
// rejects an all-absent filter with 400 validation_error.
//
// Wire shape:
//
//	{"must": [...], "should": [...], "must_not": [...]}
type QdrantFilter struct {
	Must    []FilterCondition `json:"must,omitempty"`
	Should  []FilterCondition `json:"should,omitempty"`
	MustNot []FilterCondition `json:"must_not,omitempty"`
}

// FilterCondition is a single Qdrant-style filter predicate. Exactly one of
// Match, Range, or Filter should be set alongside Key (or Key may be empty
// for nested sub-filter conditions).
type FilterCondition struct {
	Key    string        `json:"key"`
	Match  *FilterMatch  `json:"match,omitempty"`
	Range  *FilterRange  `json:"range,omitempty"`
	Filter *QdrantFilter `json:"filter,omitempty"`
}

// FilterMatch represents an exact-value or multi-value match predicate.
// Set Value for a single-value equality check, or Any for a set-membership
// ("in") check. Setting both is undefined behaviour.
type FilterMatch struct {
	Value interface{}   `json:"value,omitempty"`
	Any   []interface{} `json:"any,omitempty"`
}

// FilterRange represents a numeric range predicate. Both bounds are optional;
// omitting a bound removes that side of the range.
type FilterRange struct {
	Gte *float64 `json:"gte,omitempty"`
	Lte *float64 `json:"lte,omitempty"`
}

// IsEmpty reports true when ALL three condition arrays are empty or absent.
// An empty filter is rejected by the server with 400 validation_error; the
// typed DeleteByFilter / BulkUpdateMetadata helpers also reject it
// client-side before issuing an HTTP request.
func (f *QdrantFilter) IsEmpty() bool {
	return len(f.Must) == 0 && len(f.Should) == 0 && len(f.MustNot) == 0
}

// FilterEq returns a single-equality match condition.
//
// Example: FilterEq("topic", "index") serialises to
//
//	{"key":"topic","match":{"value":"index"}}
func FilterEq(key string, value interface{}) FilterCondition {
	return FilterCondition{
		Key:   key,
		Match: &FilterMatch{Value: value},
	}
}

// FilterIn returns a multi-value match condition ("value is one of ...").
//
// Example: FilterIn("tier", []interface{}{"hot","warm"}) serialises to
//
//	{"key":"tier","match":{"any":["hot","warm"]}}
func FilterIn(key string, values []interface{}) FilterCondition {
	return FilterCondition{
		Key:   key,
		Match: &FilterMatch{Any: values},
	}
}

// FilterRangeGteLte returns a range condition with optional lower and upper
// bounds. Pass nil for either bound to leave that side open.
//
// Example:
//
//	gte, lte := 0.5, 0.9
//	FilterRangeGteLte("score", &gte, &lte)
//
// serialises to {"key":"score","range":{"gte":0.5,"lte":0.9}}.
func FilterRangeGteLte(key string, gte, lte *float64) FilterCondition {
	return FilterCondition{
		Key:   key,
		Range: &FilterRange{Gte: gte, Lte: lte},
	}
}

// FilterNested wraps a sub-filter under a condition's filter slot.
// The key field is left empty, matching the Qdrant nested-filter convention.
//
// Example: FilterNested(MustFilter(FilterEq("k","v"))) serialises to
//
//	{"key":"","filter":{"must":[{"key":"k","match":{"value":"v"}}]}}
func FilterNested(sub QdrantFilter) FilterCondition {
	return FilterCondition{
		Filter: &sub,
	}
}

// MustFilter builds a QdrantFilter that requires ALL given conditions to be
// true (logical AND). The Should and MustNot arrays are omitted.
func MustFilter(conditions ...FilterCondition) QdrantFilter {
	return QdrantFilter{Must: conditions}
}

// ShouldFilter builds a QdrantFilter that requires AT LEAST ONE of the given
// conditions to be true (logical OR). The Must and MustNot arrays are
// omitted.
func ShouldFilter(conditions ...FilterCondition) QdrantFilter {
	return QdrantFilter{Should: conditions}
}

// MustNotFilter builds a QdrantFilter that requires ALL given conditions to
// be false (logical NOT). The Must and Should arrays are omitted.
func MustNotFilter(conditions ...FilterCondition) QdrantFilter {
	return QdrantFilter{MustNot: conditions}
}

File: sdks/python/vectorizer/filter.py
"""Typed QdrantFilter helpers for the Vectorizer Python SDK (phase23).

Matches the wire shape of the Rust + TypeScript SDKs exactly:

.. code-block:: json

    {
        "must":     [<Condition>...],
        "should":   [<Condition>...],
        "must_not": [<Condition>...]
    }

Where each ``Condition`` is ``{key, match?, range?, filter?}``:

- ``match``  — exact value (``{"value": any}``) or membership check
  (``{"any": [...]}``)
- ``range``  — numeric bounds (``{"gte": float?, "lte": float?}``)
- ``filter`` — nested ``QdrantFilter`` sub-object

Quick start::

    from vectorizer.filter import QdrantFilter, filter_eq, filter_range, must_filter

    f = must_filter(
        filter_eq("topic", "index"),
        filter_range("score", gte=0.8),
    )
    # f.to_dict() == {"must": [{"key": "topic", "match": {"value": "index"}},
    #                           {"key": "score", "range": {"gte": 0.8}}]}

"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, List, Optional


# ---------------------------------------------------------------------------
# Leaf value types
# ---------------------------------------------------------------------------


@dataclass
class FilterMatch:
    """Exact-value or membership match sub-object.

    Set exactly one of ``value`` (exact match) or ``any`` (membership):

    - ``{"value": <any>}``   — field must equal ``value``
    - ``{"any": [<any>...]}`` — field must equal any element of ``any``
    """

    value: Optional[Any] = None
    any: Optional[List[Any]] = None

    def to_dict(self) -> dict:
        """Serialise to the wire-shape dict, omitting unset fields."""
        out: dict = {}
        if self.value is not None:
            out["value"] = self.value
        if self.any is not None:
            out["any"] = list(self.any)
        return out


@dataclass
class FilterRange:
    """Numeric range bounds sub-object.

    Both bounds are optional and combined with AND semantics:

    - ``{"gte": <float>}``              — lower bound (inclusive)
    - ``{"lte": <float>}``              — upper bound (inclusive)
    - ``{"gte": <float>, "lte": <float>}`` — closed interval
    """

    gte: Optional[float] = None
    lte: Optional[float] = None

    def to_dict(self) -> dict:
        """Serialise to the wire-shape dict, omitting unset bounds."""
        out: dict = {}
        if self.gte is not None:
            out["gte"] = self.gte
        if self.lte is not None:
            out["lte"] = self.lte
        return out


# ---------------------------------------------------------------------------
# Condition
# ---------------------------------------------------------------------------


@dataclass
class FilterCondition:
    """A single filter condition attached to a payload field key.

    Exactly one of ``match``, ``range``, or ``filter`` should be set:

    - ``match``  — exact value or membership check on ``key``
    - ``range``  — numeric bounds check on ``key``
    - ``filter`` — nested compound sub-filter (``key`` is ignored by the server)

    Wire shape examples::

        {"key": "topic", "match": {"value": "index"}}
        {"key": "score", "range": {"gte": 0.5, "lte": 0.9}}
        {"key": "__nested__", "filter": {<QdrantFilter>}}
    """

    key: str
    match: Optional[FilterMatch] = None
    range: Optional[FilterRange] = None
    filter: Optional["QdrantFilter"] = None

    def to_dict(self) -> dict:
        """Serialise to the wire-shape dict, omitting unset sub-objects."""
        out: dict = {"key": self.key}
        if self.match is not None:
            out["match"] = self.match.to_dict()
        if self.range is not None:
            out["range"] = self.range.to_dict()
        if self.filter is not None:
            out["filter"] = self.filter.to_dict()
        return out


# ---------------------------------------------------------------------------
# Top-level filter
# ---------------------------------------------------------------------------


@dataclass
class QdrantFilter:
    """Top-level Qdrant-style filter accepted by ``delete_by_filter`` and
    ``bulk_update_metadata``.

    All three clause arrays are optional; omit any you don't need.  At least
    one clause with at least one condition must be present — the server rejects
    an all-absent filter with ``400 validation_error`` (message: "filter has no
    conditions").

    Wire shape::

        {"must": [...], "should": [...], "must_not": [...]}

    Use :func:`is_empty` to validate client-side before issuing the request.
    """

    must: Optional[List[FilterCondition]] = None
    should: Optional[List[FilterCondition]] = None
    must_not: Optional[List[FilterCondition]] = None

    def to_dict(self) -> dict:
        """Serialise to the wire-shape dict, omitting empty/absent clauses."""
        out: dict = {}
        if self.must:
            out["must"] = [c.to_dict() for c in self.must]
        if self.should:
            out["should"] = [c.to_dict() for c in self.should]
        if self.must_not:
            out["must_not"] = [c.to_dict() for c in self.must_not]
        return out

    def is_empty(self) -> bool:
        """Return ``True`` when none of the three clause lists has any conditions."""
        return not (self.must or self.should or self.must_not)


# ---------------------------------------------------------------------------
# Builder helpers
# ---------------------------------------------------------------------------


def filter_eq(key: str, value: Any) -> FilterCondition:
    """Build an exact-match condition.

    Wire: ``{"key": "<key>", "match": {"value": <value>}}``

    Args:
        key:   Payload field path (dot-separated for nested fields).
        value: Value the field must equal.
    """
    return FilterCondition(key=key, match=FilterMatch(value=value))


def filter_in(key: str, values: List[Any]) -> FilterCondition:
    """Build a multi-value membership condition (field IN [...]).

    Wire: ``{"key": "<key>", "match": {"any": [...]}}``

    Args:
        key:    Payload field path.
        values: List of values; the field must equal any one of them.
    """
    return FilterCondition(key=key, match=FilterMatch(any=list(values)))


def filter_range(
    key: str,
    gte: Optional[float] = None,
    lte: Optional[float] = None,
) -> FilterCondition:
    """Build a numeric range condition.

    Wire: ``{"key": "<key>", "range": {"gte": <num>?, "lte": <num>?}}``

    Args:
        key: Payload field path.
        gte: Lower bound (inclusive).  Omit to leave unbounded.
        lte: Upper bound (inclusive).  Omit to leave unbounded.
    """
    return FilterCondition(key=key, range=FilterRange(gte=gte, lte=lte))


def filter_nested(sub: QdrantFilter) -> FilterCondition:
    """Wrap a sub-filter as a nested condition.

    Wire: ``{"key": "__nested__", "filter": <QdrantFilter>}``

    Args:
        sub: Nested compound filter to embed.
    """
    return FilterCondition(key="__nested__", filter=sub)


def must_filter(*conditions: FilterCondition) -> QdrantFilter:
    """Build a filter requiring ALL given conditions to be true (AND semantics).

    Wire: ``{"must": [...]}``

    Args:
        *conditions: One or more :class:`FilterCondition` instances.
    """
    return QdrantFilter(must=list(conditions))


def should_filter(*conditions: FilterCondition) -> QdrantFilter:
    """Build a filter requiring AT LEAST ONE condition to be true (OR semantics).

    Wire: ``{"should": [...]}``

    Args:
        *conditions: One or more :class:`FilterCondition` instances.
    """
    return QdrantFilter(should=list(conditions))


def must_not_filter(*conditions: FilterCondition) -> QdrantFilter:
    """Build a filter requiring ALL conditions to be false (NOT semantics).

    Wire: ``{"must_not": [...]}``

    Args:
        *conditions: One or more :class:`FilterCondition` instances.
    """
    return QdrantFilter(must_not=list(conditions))
